### PR TITLE
Development: restructure the repository

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
This PR aims to reconfigure the repository's folder structure and introduce a basic CI that runs tests and linters.

Changes Made:
- Added pyproject.toml for installing.
- Removed setup.py.
- Added tox.ini file for facilitating testing/linting on multiple OS.
- Added .github/workflows/ci.yml for basic CI.
- Added .github/workflows/connect.yml to SSH to GitHub runners for debugging when manually triggered.
- Set all of the developers as code owners (globally for now).
- Moved the code to src/neurostatslib.

The basis.py code will be substantially rewritten, so the differences between this development version and the main are not important. Once this PR goes through, I will start following the git-flow, creating a branch from development that will be used to restructure the basis function class.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1 @@
+@BalzaniEdoardo @billbrod @ahwillia @gviejo
```
Comment (billbrod) [inline .github/CODEOWNERS:1]: Confused about this codeowners file: my understanding is that these files should be structured like `path @user`, where `user` is then the owner for every file at `path` (as in [the docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)). Thus, I was expected this to be `* @BalzaniEdoardo`, meaning Edoardo needs to approve every PR to this repo. I think there's something weird here where it thinks Edoardo is the path. Also, in general, I thought we had agreed to have one owner per repo (else, we'd all have to review every PR), potentially making different people "subowners" (e.g., making me the owner of `ci.yml` or `.readthedocs.yml` or Guillaume the owner of `tox.ini`)

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+name: ci
+
+on:
+#  push:
+#    branches:
+#      - folder_struct_and_setup
```
Comment (billbrod) [inline .github/workflows/ci.yml:6]: delete these if not using them

File: .github/workflows/connect.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+name: connect
```
Comment (billbrod) [inline .github/workflows/connect.yml:1]: add a comment here to describe what this workflow is for and when you'd use it? (same for `ci.yml`)

File: src/neurostatslib/basis.py [hunk 4]
```diff
@@ -0,0 +1,621 @@
+"""Bases classes.
```
Comment (billbrod) [inline src/neurostatslib/basis.py:1]: why is this file double the length of the old `basis.py`? looks like it contains things that are in the `edoardo` branch -- the new spline bases. those should be removed so that we keep the two changes separate.

File: src/neurostatslib/glm.py [hunk 5]
```diff
@@ -1,13 +1,17 @@
+"""GLM core module
```
Comment (billbrod) [inline src/neurostatslib/glm.py:1]: are the only differences here the result of running the linters?

File: src/neurostatslib/utils.py [hunk 6]
```diff
@@ -1,26 +1,29 @@
+"""Utility functions for data pre-processing
```
Comment (billbrod) [inline src/neurostatslib/utils.py:1]: are the only changes here the result of linters?

File: tests/test_BSpline.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# content of ./test_parametrize.py
+import pytest
```
Comment (billbrod) [inline tests/test_BSpline.py:2]: should be removed from this PR for now

File: tests/test_glm_synthetic.py [hunk 8]
```diff
@@ -49,26 +49,26 @@
...
 fit_pred = fitted_model.predict(spike_data)
+#
+# fig, axes = plt.subplots(2, 1)
+# axes[0].plot(onp.arange(nt), spike_data[0])
+# axes[0].plot(onp.arange(ws, nt + 1), sim_pred[0])
+# axes[0].plot(onp.arange(ws, nt + 1), fit_pred[0])
+# axes[1].plot(onp.arange(nt), spike_data[1])
+# axes[1].plot(onp.arange(ws, nt + 1), sim_pred[1])
+# axes[1].plot(onp.arange(ws, nt + 1), fit_pred[1])
+# plt.show()
```
Comment (billbrod) [inline tests/test_glm_synthetic.py:60]: why are these lines commented out?

File: tests/test_glm_synthetic.py [hunk 9]
```diff
@@ -16,7 +16,7 @@
     n_basis_funcs=5,
     window_size=ws
 )
-sim_pts = nsl.sample_points.raised_cosine_log(5, ws)
+sim_pts = jnp.clip(neurostatslib.sample_points.raised_cosine_log(5, ws),0,100)
```
Comment (billbrod) [inline tests/test_glm_synthetic.py:19]: this shouldn't be necessary with the change in `raised_cosine_log`, right?

File: tests/test_glm_synthetic_single_neuron.py [hunk 10]
```diff
@@ -15,7 +15,7 @@
     n_basis_funcs=5,
     window_size=ws
 )
-sim_pts = nsl.sample_points.raised_cosine_log(5, ws)
+sim_pts = jnp.clip(neurostatslib.sample_points.raised_cosine_log(5, ws),0,100)
```
Comment (billbrod) [inline tests/test_glm_synthetic_single_neuron.py:18]: same comments as `test_glm_synthetic`:
 - why are the lines commented out?
 - this clip shouldn't be necessary, right?

File: tests/test_glm_synthetic_single_neuron.py [hunk 11]
```diff
@@ -1,12 +1,12 @@
 import jax
 import jax.numpy as jnp
 import numpy as onp
-import neurostatslib as nsl
 from neurostatslib.glm import GLM
 from neurostatslib.basis import RaisedCosineBasis
 import matplotlib.pyplot as plt
 jax.config.update("jax_platform_name", "cpu")
 jax.config.update("jax_enable_x64", True)
+import neurostatslib
```
Comment (billbrod) [inline tests/test_glm_synthetic_single_neuron.py:9]: what's the matter with `import neurostatslib as nsl`?

File: tests/test_glm_synthetic_single_neuron.py [hunk 12]
```diff
@@ -1,12 +1,12 @@
 import jax
```
Comment (billbrod) [inline tests/test_glm_synthetic_single_neuron.py:1]: also, this script is basically the same as `test_glm_synthetic.py`. I'm thinking maybe we combine them: make a function that runs this, with an option to set the number of neurons to include. but that doesn't belong in this PR

**Review by billbrod — CHANGES_REQUESTED** [review]
In addition to my comments, I don't understand the structure of `pyproject.toml` and `tox.ini`. If we can document why we included the sections we did and what they mean in the template repo, I'm fine, but just want to highlight the same issues here.

File: .github/CODEOWNERS (hunk 1 above)
Comment (BalzaniEdoardo) [reply .github/CODEOWNERS:1]: with this syntax, I am setting up everyone as code owner. I suggested in the comments that we can make it more specific later but since we did not discuss the specifics yet I went for this temporary solution. If you prefer I can set myself as the only owner for now or add you guys as specific script owner as you are suggesting

File: src/neurostatslib/basis.py (hunk 4 above)
Comment (BalzaniEdoardo) [reply src/neurostatslib/basis.py:1]: I mention this in the PR request comment, the basis class will be rebuild from scratch, so ignore any diffs. (this version works, doesn't brake the code but still will be re-structured)

File: src/neurostatslib/glm.py (hunk 5 above)
Comment (BalzaniEdoardo) [reply src/neurostatslib/glm.py:1]: yes

File: tests/test_glm_synthetic.py (hunk 8 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic.py:60]: all this test code will need to be re-written. I'll move visualization scripts into the scripts we will use for the documentation

File: tests/test_glm_synthetic.py (hunk 9 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic.py:19]: now yes, anyways, this code which is more for visualization will be moved to a dedicated documentation script.

File: tests/test_glm_synthetic_single_neuron.py (hunk 10 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic_single_neuron.py:18]: as I said before, this will be moved

File: tests/test_glm_synthetic_single_neuron.py (hunk 11 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic_single_neuron.py:9]: sounds good

File: src/neurostatslib/basis.py (hunk 4 above)
Comment (BalzaniEdoardo) [reply src/neurostatslib/basis.py]: reverted to the main and created a specific branch to tackle the basis function issue

File: tests/test_glm_synthetic.py (hunk 8 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic.py]: reverted tests to the main version. tests will be reformatted when the functions that they are pointing to will be modified

File: tests/test_glm_synthetic_single_neuron.py (hunk 11 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic_single_neuron.py]: for now I kept it consistent with the main

File: tests/test_glm_synthetic_single_neuron.py (hunk 12 above)
Comment (BalzaniEdoardo) [reply tests/test_glm_synthetic_single_neuron.py]: same as above, I kept it consistent with jax

File: .github/CODEOWNERS (hunk 1 above)
Comment (BalzaniEdoardo) [reply .github/CODEOWNERS:1]: edited, for now I set myself as codeowner, we can be more specific  once we agree on who does what on this repo

File: .github/workflows/connect.yml (hunk 3 above)
Comment (BalzaniEdoardo) [reply .github/workflows/connect.yml:1]: done
